src/Forms/GridField/GridFieldFilterHeader.php

```diff
@@ -15,10 +22,150 @@
  * GridFieldFilterHeader alters the {@link GridField} with some filtering
  * fields in the header of each column.
  *
+ * filter header with customisable filter fields
+ *
+ * REQUIREMENTS:
+ *
+ * table header component needs to be present in the GridField (for example GridFieldSortableHeader)
+ * the last column of the table needs to have a vacant header so the filter widget could be displayed there
+ * for example you can't have the last column with sorting header and filter widget at the same time
+ *
+ * ALIAS FIELDS
+ *
+ * in the case that the column name of the table is different from matching field name via the summary_fields
+ * a field alias needs to be specified
+ *
+ * configuration format:
+ *
+ * 'field_alias' => 'field_name'
+ *
+ * for example if summary fields looks like this:
+ *
+ * private static $summary_fields = [
+ *   'getTitleSummary' => 'Title',
+ *   'City.Name' => 'City',
+ *   'Expires.Nice' => 'Expires',
+ * ];
+ *
+ * alias fields need to be specified to match the summary fields mapping
+ *
+ * ->setAliasFields([
+ *   'getTitleSummary' => 'Title',
+ *   'City.Name' => 'City',
+ *   'Expires.Nice' => 'Expires',
+ * ])
+ *
+ * Field alias is always applied first so there is no need to use the original name in other configuration
+ *
+ * CUSTOM FIELDS
+ *
+ * TextField is used as a default field for filters however this may be too crude in some situations
+ * Custom fields are used to override the default fields with any field type that is needed
+ *
+ * configuration format:
+ *
+ * 'field_name' => 'field_object'
+ *
+ * for example if we want to use Dropdown field configuration below has to be used:
+ *
+ * ->setCustomFields([
+ *  'City' => DropdownField::create('', '', $cities),
+ * ])
+ *
+ * Note that Title and Name of the field are left empty as they are not used (Title) or are auto-populated (Name)
+ *
+ * CUSTOM FILTERS
+ *
+ * Partial match filter is used by default, but in some situations this can't be used as other filters are required
+ *
+ * configuration format:
+ *
+ * 'field_name' => 'filter_specification' (Closure or string)
+ *
+ * This component comes with Partial filter, Exact filter and Relation filter to cover most common cases
+ *
+ * for example we may want to use Exact match filter as we are filtering by IDs
+ *
+ * ->setCustomFilters([
+ *   'MediaType' => GridFieldFilterHeader::FILTER_EXACT,
+ *   'City' => function (Filterable $list, $columnName, $value) {
+ *     return $list->filter('CityID:ExactMatch', $value);
+ *   },
+ * ])
+ *
+ * custom filter can be either a string which identifies one of the filters that are available
+ * or a custom Closure which is expected to filter the list
+ * this is a great way to cover edge cases as the implementation of the filter is up to the developer
```

Comment:
A lot of the functionality provided by this change seems to be covered by SearchContext. Have you considered using an injectable context instead of replicating this functionality?

This includes:

 - Scaffolding form fields
 - Querying results from a search
 - Handling http requests